File: src/domain/tests/properties.test.ts
// Свойства движка на случайных журналах (детерминированный seed).
// Эталон — поминутная модель: каждая минута журнала имеет один режим,
// правила регламента применяются к сериям минут, без блоков и склейки.

import { describe, expect, it } from 'vitest';
import type { ComplianceMetrics } from '../compliance';
import { buildJournal } from '../journal';
import { LIMITS } from '../limits';
import { WEEK, weekStartUtc } from '../time';
import type { ActivityEntry, ActivityType } from '../types';
import { MINUTE, calc, logFrom, rng, utc, type Seg } from './helpers';

const REST_DURATIONS = [
  1, 10, 14, 15, 16, 20, 29, 30, 31, 44, 45, 46, 60, 120, 179, 180, 181, 300, 539, 540, 541, 600, 659, 660, 661,
  720, 1439, 1440, 1441, 2000, 2699, 2700, 3000,
];
const WORK_DURATIONS = [1, 5, 15, 30, 45, 60, 90, 120, 179, 180, 240, 269, 270, 271, 300, 400];
const ACTIVITIES: ActivityType[] = ['DRIVE', 'DRIVE', 'DRIVE', 'REST', 'REST', 'WORK', 'POA'];

function randomLog(seed: number): { entries: ActivityEntry[]; now: number } {
  const r = rng(seed);
  const start = utc('2026-08-31 00:00') + r.int(0, 21 * 24 * 60) * MINUTE;
  const segs: Seg[] = Array.from({ length: r.int(1, 60) }, () => {
    const activity = r.pick(ACTIVITIES);
    return [activity, r.pick(activity === 'REST' ? REST_DURATIONS : WORK_DURATIONS)];
  });
  return logFrom(start, segs);
}

interface Run {
  from: number;
  to: number;
}

/** Поминутная эталонная модель ст. 6–8 для журнала без парома и «конца дня». */
function oracle(entries: ActivityEntry[], now: number) {
  const origin = entries[0].startTime;
  const n = (now - origin) / MINUTE;
  const act: ActivityType[] = new Array(n);
  for (const e of entries) {
    const to = ((e.endTime ?? now) - origin) / MINUTE;
    for (let t = (e.startTime - origin) / MINUTE; t < to; t++) act[t] = e.activity;
  }
  const len = (r: Run) => r.to - r.from;

  const rests: Run[] = [];
  for (let t = 0; t < n; t++) {
    if (act[t] !== 'REST') continue;
    const last = rests[rests.length - 1];
    if (last && last.to === t) last.to = t + 1;
    else rests.push({ from: t, to: t + 1 });
  }
  const closed = (r: Run) => r.to < n;

  // Смены — участки между отдыхами ≥ 9 ч, в которых есть что-то кроме отдыха
  const shifts: { from: number; to: number; restAfter: Run | null }[] = [];
  let cursor = 0;
  for (const r of [...rests.filter((x) => len(x) >= LIMITS.dailyRestReduced), null]) {
    const end = r ? r.from : n;
    let first = cursor;
    while (first < end && act[first] === 'REST') first++;
    if (first < end) shifts.push({ from: first, to: end, restAfter: r });
    cursor = r ? r.to : n;
  }
  const last = shifts[shifts.length - 1];
  const current = last && last.restAfter === null ? last : null;

  // Ст. 7: серия отдыха ≥ 45 мин или 15, затем ≥ 30
  let continuous = 0;
  let daily = 0;
  if (current) {
    let firstPart = false;
    let run = 0;
    const closeRun = () => {
      if (run >= 45 || (firstPart && run >= 30)) {
        continuous = 0;
        firstPart = false;
      } else if (!firstPart && run >= 15) firstPart = true;
      run = 0;
    };
    for (let t = current.from; t < n; t++) {
      if (act[t] === 'REST') {
        run++;
        continue;
      }
      if (run) closeRun();
      if (act[t] === 'DRIVE') {
        continuous++;
        daily++;
      }
    }
    if (run) closeRun();
  }

  const lastRun = rests[rests.length - 1];
  const resting = act[n - 1] === 'REST';
  const shiftMinutes = current ? (resting ? lastRun.from : n) - current.from : 0;

  const driveSince = (ts: number) => {
    let total = 0;
    for (let t = Math.max(0, (ts - origin) / MINUTE); t < n; t++) if (act[t] === 'DRIVE') total++;
    return total;
  };
  const weekStart = weekStartUtc(now);

  // Ст. 8(2) и 8(4): сокращённые отдыхи после последнего завершённого недельного;
  // статус — по части отдыха в окне 24 ч от начала смены
  const weekly = rests.filter((r) => len(r) >= LIMITS.weeklyRestReduced && closed(r));
  const since = weekly.length ? weekly[weekly.length - 1].to : -Infinity;
  let reducedRests = 0;
  for (const s of shifts) {
    const r = s.restAfter;
    if (!r || !closed(r) || r.from < since || len(r) >= LIMITS.weeklyRestReduced) continue;
    const inWindow = Math.min(len(r), s.from + LIMITS.shiftWindowSolo - r.from);
    const split = rests.some((x) => x.from >= s.from && x.to <= s.to && len(x) >= LIMITS.dailyRestSplitFirst);
    if (inWindow >= LIMITS.dailyRestReduced && inWindow < LIMITS.dailyRestRegular && !split) reducedRests++;
  }

  return {
    shifts: shifts.length,
    shiftStart: current ? origin + current.from * MINUTE : null,
    shiftMinutes,
    continuous,
    daily,
    weekly: driveSince(weekStart),
    fortnight: driveSince(weekStart - WEEK),
    offDutyRest: !current && resting && len(lastRun) >= LIMITS.dailyRestReduced ? len(lastRun) : null,
    reducedRests,
  };
}

/** Всё, что видит водитель, без идентификаторов записей. */
function visible(m: ComplianceMetrics) {
  const { timeline, infringements, ...rest } = m;
  return {
    ...rest,
    shift: m.shift && { start: m.shift.start, drive: m.shift.driveMinutes, work: m.shift.workMinutes },
    shifts: timeline.shifts.map((s) => [s.start, s.end, s.driveMinutes, s.workMinutes, s.poaMinutes, s.breakMinutes]),
    infringements: infringements.map((i) => [i.key, i.params]),
  };
}

const SEEDS = Array.from({ length: 400 }, (_, i) => i + 1);

describe('свойства на случайных журналах', () => {
  it('совпадают с поминутной эталонной моделью', () => {
    for (const seed of SEEDS) {
      const { entries, now } = randomLog(seed);
      const m = calc(entries, now);
      const o = oracle(entries, now);
      const actual = {
        shifts: m.timeline.shifts.length,
        shiftStart: m.shift?.start ?? null,
        shiftMinutes: m.shiftMinutes,
        continuous: m.continuousDriveMinutes,
        daily: m.dailyDriveMinutes,
        weekly: m.weeklyDriveMinutes,
        fortnight: m.fortnightDriveMinutes,
        offDutyRest: m.offDutyRest?.minutes ?? null,
        reducedRests: m.reducedRestsUsed,
      };
      expect(actual, `seed ${seed}`).toEqual(o);
    }
  });

  it('повторное нажатие того же режима (дробление записи) ничего не меняет', () => {
    for (const seed of SEEDS) {
      const { entries, now } = randomLog(seed);
      const r = rng(seed * 7919);
      const i = r.int(0, entries.length - 1);
      const e = entries[i];
      const end = e.endTime ?? now;
      if (end - e.startTime < 2 * MINUTE) continue;
      const cut = e.startTime + r.int(1, (end - e.startTime) / MINUTE - 1) * MINUTE;
      const split = [
        ...entries.slice(0, i),
        { ...e, endTime: cut },
        { ...e, id: `${e.id}-b`, startTime: cut },
        ...entries.slice(i + 1),
      ];
      expect(visible(calc(split, now)), `seed ${seed}`).toEqual(visible(calc(entries, now)));
    }
  });

  it('порядок записей в хранилище не важен', () => {
    for (const seed of SEEDS.slice(0, 100)) {
      const { entries, now } = randomLog(seed);
      const r = rng(seed * 104729);
      const shuffled = [...entries].sort(() => r.next() - 0.5);
      expect(visible(calc(shuffled, now)), `seed ${seed}`).toEqual(visible(calc(entries, now)));
    }
  });

  it('время смены раскладывается на режимы без остатка', () => {
    for (const seed of SEEDS) {
      const { entries, now } = randomLog(seed);
      for (const s of calc(entries, now).timeline.shifts) {
        const span = ((s.end ?? now) - s.start) / MINUTE;
        expect(s.driveMinutes + s.workMinutes + s.poaMinutes + s.breakMinutes, `seed ${seed}`).toBe(span);
      }
    }
  });

  it('всё вождение журнала распределено по сменам', () => {
    for (const seed of SEEDS) {
      const { entries, now } = randomLog(seed);
      const total = entries
        .filter((e) => e.activity === 'DRIVE')
        .reduce((sum, e) => sum + ((e.endTime ?? now) - e.startTime) / MINUTE, 0);
      const byShifts = calc(entries, now).timeline.shifts.reduce((sum, s) => sum + s.driveMinutes, 0);
      expect(byShifts, `seed ${seed}`).toBe(total);
    }
  });

  it('журнал и главный экран считают неделю одинаково', () => {
    for (const seed of SEEDS) {
      const { entries, now } = randomLog(seed);
      const m = calc(entries, now);
      const week = buildJournal({ timeline: m.timeline, manualShifts: [], meta: {}, crewMode: 'SOLO', now }).find(
        (w) => w.isCurrent,
      )!;
      expect([week.driveMinutes, week.fortnightMinutes], `seed ${seed}`).toEqual([
        m.weeklyDriveMinutes,
        m.fortnightDriveMinutes,
      ]);
    }
  });

  it('остатки и нарушения согласованы с посчитанными значениями', () => {
    for (const seed of SEEDS) {
      const { entries, now } = randomLog(seed);
      const m = calc(entries, now);
      const has = (key: string) => m.infringements.some((i) => i.key === key);
      const msg = `seed ${seed}`;
      expect(m.driveUntilBreakMinutes, msg).toBe(Math.max(0, LIMITS.continuousDrive - m.continuousDriveMinutes));
      expect(m.weeklyDriveRemainingMinutes, msg).toBeGreaterThanOrEqual(0);
      expect(m.weeklyDriveRemainingMinutes, msg).toBeLessThanOrEqual(LIMITS.weeklyDrive);
      expect(m.extensionsLeft, msg).toBe(Math.max(0, 2 - m.extensionsUsed));
      expect(m.reducedRestsLeft, msg).toBe(Math.max(0, 3 - m.reducedRestsUsed));
      expect(m.continuousDriveMinutes, msg).toBeLessThanOrEqual(m.dailyDriveMinutes);
      expect(has('continuousExceeded'), msg).toBe(m.continuousDriveMinutes > LIMITS.continuousDrive);
      expect(has('dailyDriveExceeded'), msg).toBe(m.dailyDriveMinutes > m.dailyDriveLimitMinutes);
      expect(has('weeklyDriveExceeded'), msg).toBe(m.weeklyDriveMinutes > LIMITS.weeklyDrive);
      expect(has('fortnightDriveExceeded'), msg).toBe(m.fortnightDriveMinutes > LIMITS.fortnightDrive);
      expect(has('shiftExceeded'), msg).toBe(m.shift !== null && m.shiftMinutes > m.shiftLimitMinutes);
    }
  });
});

File: packages/tacho_engine/test/properties_test.dart
// Свойства движка на случайных журналах (детерминированный seed).
// Эталон — поминутная модель: каждая минута журнала имеет один режим,
// правила регламента применяются к сериям минут, без блоков и склейки.

import 'dart:math';

import 'package:tacho_engine/tacho_engine.dart';
import 'package:test/test.dart';

import 'helpers.dart';

const restDurations = [
  1, 10, 14, 15, 16, 20, 29, 30, 31, 44, 45, 46, 60, 120, 179, 180, 181, //
  300, 539, 540, 541, 600, 659, 660, 661, 720, 1439, 1440, 1441, 2000, //
  2699, 2700, 3000,
];
const workDurations = [
  1,
  5,
  15,
  30,
  45,
  60,
  90,
  120,
  179,
  180,
  240,
  269,
  270,
  271,
  300,
  400,
];
const List<DriverMode> activities = [
  DriverMode.driving,
  DriverMode.driving,
  DriverMode.driving,
  DriverMode.rest,
  DriverMode.rest,
  DriverMode.otherWork,
  DriverMode.availability,
];

({List<ActivityPeriod> periods, DateTime now}) randomLog(int seed) {
  final r = Rng(seed);
  final start = utc('2026-08-31 00:00')
      .add(minutes(r.nextInt(0, 21 * 24 * 60)));
  final segs = [for (var i = r.nextInt(1, 60); i > 0; i--) _randomSeg(r)];
  return logFrom(start, segs);
}

Seg _randomSeg(Rng r) {
  final mode = r.pick(activities);
  final d = r.pick(mode == DriverMode.rest ? restDurations : workDurations);
  return Seg(mode, minutes(d));
}

/// Серия минут [from, to) от начала журнала.
class _Run {
  new(this.from, this.to);

  final int from;
  int to;

  int get length => to - from;
}

/// Поминутная эталонная модель ст. 6–8 для журнала без парома и «конца дня».
Map<String, Object?> oracle(List<ActivityPeriod> periods, DateTime now) {
  final origin = periods.first.start;
  int offset(DateTime t) => t.difference(origin).inMinutes;
  final n = offset(now);
  final act = List<DriverMode?>.filled(n, null);
  for (final p in periods) {
    for (var t = offset(p.start); t < offset(p.end ?? now); t++) {
      act[t] = p.mode;
    }
  }

  final rests = <_Run>[];
  for (var t = 0; t < n; t++) {
    if (act[t] != DriverMode.rest) continue;
    if (rests.isNotEmpty && rests.last.to == t) {
      rests.last.to = t + 1;
    } else {
      rests.add(_Run(t, t + 1));
    }
  }
  bool closed(_Run r) => r.to < n;
  final reduced = EuLimits.dailyRestReduced.inMinutes;

  // Смены — участки между отдыхами ≥ 9 ч, в которых есть что-то кроме отдыха
  final shifts = <({int from, int to, _Run? restAfter})>[];
  var cursor = 0;
  for (final r in [...rests.where((x) => x.length >= reduced), null]) {
    final end = r?.from ?? n;
    var first = cursor;
    while (first < end && act[first] == DriverMode.rest) {
      first++;
    }
    if (first < end) shifts.add((from: first, to: end, restAfter: r));
    cursor = r?.to ?? n;
  }
  final current = shifts.isNotEmpty && shifts.last.restAfter == null
      ? shifts.last
      : null;

  // Ст. 7: серия отдыха ≥ 45 мин или 15, затем ≥ 30
  var continuous = 0;
  var daily = 0;
  if (current != null) {
    var firstPart = false;
    var run = 0;
    void closeRun() {
      if (run >= 45 || (firstPart && run >= 30)) {
        continuous = 0;
        firstPart = false;
      } else if (!firstPart && run >= 15) {
        firstPart = true;
      }
      run = 0;
    }

    for (var t = current.from; t < n; t++) {
      if (act[t] == DriverMode.rest) {
        run++;
        continue;
      }
      if (run > 0) closeRun();
      if (act[t] == DriverMode.driving) {
        continuous++;
        daily++;
      }
    }
    if (run > 0) closeRun();
  }

  final lastRun = rests.isEmpty ? null : rests.last;
  final resting = act[n - 1] == DriverMode.rest;
  final shiftMinutes = current == null
      ? 0
      : (resting ? lastRun!.from : n) - current.from;

  int drivingSince(DateTime ts) {
    var total = 0;
    final from = offset(ts);
    for (var t = from < 0 ? 0 : from; t < n; t++) {
      if (act[t] == DriverMode.driving) total++;
    }
    return total;
  }

  final weekStart = weekStartUtc(now);

  // Ст. 8(2) и 8(4): сокращённые отдыхи после последнего завершённого
  // недельного; статус — по части отдыха в окне 24 ч от начала смены
  final weekly = rests
      .where(
        (r) => r.length >= EuLimits.weeklyRestReduced.inMinutes && closed(r),
      )
      .toList();
  final since = weekly.isEmpty ? null : weekly.last.to;
  var reducedRests = 0;
  for (final s in shifts) {
    final r = s.restAfter;
    if (r == null ||
        !closed(r) ||
        (since != null && r.from < since) ||
        r.length >= EuLimits.weeklyRestReduced.inMinutes) {
      continue;
    }
    final window = s.from + EuLimits.workdayWindow.inMinutes - r.from;
    final inWindow = r.length < window ? r.length : window;
    final split = rests.any(
      (x) =>
          x.from >= s.from &&
          x.to <= s.to &&
          x.length >= EuLimits.dailyRestSplitFirst.inMinutes,
    );
    if (inWindow >= reduced &&
        inWindow < EuLimits.dailyRestRegular.inMinutes &&
        !split) {
      reducedRests++;
    }
  }

  return {
    'shifts': shifts.length,
    'shiftStart': current == null ? null : origin.add(minutes(current.from)),
    'shiftMinutes': shiftMinutes,
    'continuous': continuous,
    'daily': daily,
    'weekly': drivingSince(weekStart),
    'fortnight': drivingSince(weekStart.subtract(week)),
    'offDutyRest': current == null && resting && lastRun!.length >= reduced
        ? lastRun.length
        : null,
    'reducedRests': reducedRests,
  };
}

/// Всё, что видит водитель, без идентификаторов записей.
Map<String, Object?> visible(ComplianceSnapshot m) => {
  'currentMode': m.currentMode,
  'currentModeStart': m.currentModeStart,
  'currentModeDuration': m.currentModeDuration,
  'status': m.status,
  'shift': m.shift == null
      ? null
      : [m.shift!.start, m.shift!.driving, m.shift!.otherWork],
  'shiftDuration': m.shiftDuration,
  'shiftLimit': m.shiftLimit,
  'dailyRestDeadline': m.dailyRestDeadline,
  'offDutyRest': m.offDutyRest == null
      ? null
      : [m.offDutyRest!.start, m.offDutyRest!.duration, m.offDutyRest!.weekly],
  'continuousDriving': m.continuousDriving,
  'breakFirstPart': m.breakFirstPart,
  'breakRequired': m.breakRequired,
  'currentBreak': m.currentBreak == null
      ? null
      : [m.currentBreak!.start, m.currentBreak!.duration],
  'dailyDriving': m.dailyDriving,
  'dailyDrivingLimit': m.dailyDrivingLimit,
  'extensionsUsed': m.extensionsUsed,
  'reducedRestsUsed': m.reducedRestsUsed,
  'weeklyDriving': m.weeklyDriving,
  'fortnightDriving': m.fortnightDriving,
  'lastWeeklyRest': m.lastWeeklyRest,
  'previousWeeklyRest': m.previousWeeklyRest,
  'workWeekDuration': m.workWeekDuration,
  'compensation': m.compensation,
  'shifts': [
    for (final s in m.timeline.shifts)
      [s.start, s.end, s.driving, s.otherWork, s.availability, s.breaks],
  ],
  'infringements': m.infringements,
};

final List<int> seeds = [for (var i = 1; i <= 400; i++) i];

void main() {
  group('свойства на случайных журналах', () {
    test('совпадают с поминутной эталонной моделью', () {
      for (final seed in seeds) {
        final log = randomLog(seed);
        final m = calc(log.periods, log.now);
        final actual = {
          'shifts': m.timeline.shifts.length,
          'shiftStart': m.shift?.start,
          'shiftMinutes': m.shiftDuration.inMinutes,
          'continuous': m.continuousDriving.inMinutes,
          'daily': m.dailyDriving.inMinutes,
          'weekly': m.weeklyDriving.inMinutes,
          'fortnight': m.fortnightDriving.inMinutes,
          'offDutyRest': m.offDutyRest?.duration.inMinutes,
          'reducedRests': m.reducedRestsUsed,
        };
        expect(actual, oracle(log.periods, log.now), reason: 'seed $seed');
      }
    });

    test('повторное нажатие того же режима (дробление записи) ничего не '
        'меняет', () {
      for (final seed in seeds) {
        final log = randomLog(seed);
        final periods = log.periods;
        final r = Rng(seed * 7919);
        final i = r.nextInt(0, periods.length - 1);
        final p = periods[i];
        final length = p.durationAt(log.now).inMinutes;
        if (length < 2) continue;
        final cut = p.start.add(minutes(r.nextInt(1, length - 1)));
        final split = [
          ...periods.sublist(0, i),
          p.withEnd(cut),
          ActivityPeriod(mode: p.mode, start: cut, end: p.end),
          ...periods.sublist(i + 1),
        ];
        expect(
          visible(calc(split, log.now)),
          visible(calc(periods, log.now)),
          reason: 'seed $seed',
        );
      }
    });

    test('порядок записей в хранилище не важен', () {
      for (final seed in seeds.take(100)) {
        final log = randomLog(seed);
        final shuffled = [...log.periods]..shuffle(Random(seed));
        expect(
          visible(calc(shuffled, log.now)),
          visible(calc(log.periods, log.now)),
          reason: 'seed $seed',
        );
      }
    });

    test('время смены раскладывается на режимы без остатка', () {
      for (final seed in seeds) {
        final log = randomLog(seed);
        for (final s in calc(log.periods, log.now).timeline.shifts) {
          expect(
            s.driving + s.otherWork + s.availability + s.breaks,
            (s.end ?? log.now).difference(s.start),
            reason: 'seed $seed',
          );
        }
      }
    });

    test('всё вождение журнала распределено по сменам', () {
      for (final seed in seeds) {
        final log = randomLog(seed);
        final total = log.periods
            .where((p) => p.mode == DriverMode.driving)
            .fold(Duration.zero, (sum, p) => sum + p.durationAt(log.now));
        final byShifts = calc(
          log.periods,
          log.now,
        ).timeline.shifts.fold(Duration.zero, (sum, s) => sum + s.driving);
        expect(byShifts, total, reason: 'seed $seed');
      }
    });

    test('журнал и главный экран считают неделю одинаково', () {
      for (final seed in seeds) {
        final log = randomLog(seed);
        final m = calc(log.periods, log.now);
        final week = buildJournal(
          timeline: m.timeline,
          now: log.now,
        ).firstWhere((w) => w.isCurrent);
        expect(
          [week.driving, week.fortnightDriving],
          [m.weeklyDriving, m.fortnightDriving],
          reason: 'seed $seed',
        );
      }
    });

    test('остатки и нарушения согласованы с посчитанными значениями', () {
      for (final seed in seeds) {
        final log = randomLog(seed);
        final m = calc(log.periods, log.now);
        final msg = 'seed $seed';
        Duration clamp(Duration d) => d.isNegative ? Duration.zero : d;
        expect(
          m.drivingUntilBreak,
          clamp(EuLimits.continuousDriving - m.continuousDriving),
          reason: msg,
        );
        expect(m.weeklyDrivingRemaining.isNegative, isFalse, reason: msg);
        expect(
          m.weeklyDrivingRemaining <= EuLimits.weeklyDriving,
          isTrue,
          reason: msg,
        );
        expect(
          m.extensionsLeft,
          (2 - m.extensionsUsed).clamp(0, 2),
          reason: msg,
        );
        expect(
          m.reducedRestsLeft,
          (3 - m.reducedRestsUsed).clamp(0, 3),
          reason: msg,
        );
        expect(m.continuousDriving <= m.dailyDriving, isTrue, reason: msg);
        expect(
          m.has(InfringementType.continuousExceeded),
          m.continuousDriving > EuLimits.continuousDriving,
          reason: msg,
        );
        expect(
          m.has(InfringementType.dailyDriveExceeded),
          m.dailyDriving > m.dailyDrivingLimit,
          reason: msg,
        );
        expect(
          m.has(InfringementType.weeklyDriveExceeded),
          m.weeklyDriving > EuLimits.weeklyDriving,
          reason: msg,
        );
        expect(
          m.has(InfringementType.fortnightDriveExceeded),
          m.fortnightDriving > EuLimits.fortnightDriving,
          reason: msg,
        );
        expect(
          m.has(InfringementType.shiftExceeded),
          m.shift != null && m.shiftDuration > m.shiftLimit,
          reason: msg,
        );
      }
    });
  });
}
